activation/activation.go

```diff
@@ -346,35 +348,143 @@ func (b *Builder) buildInitialPost(ctx context.Context, nodeID types.NodeID) err
-		if err := nipost.RemoveInitialPost(b.localDB, nodeID); err != nil {
-			b.log.Fatal("failed to remove initial post", log.ZShortStringer("smesherID", nodeID), zap.Error(err))
-		}
-		return fmt.Errorf("initial POST is invalid: %w", err)
+		return nil, nil, fmt.Errorf("initial POST is invalid: %w", err)
+	}
+
+	if err := nipost.AddPost(b.localDB, nodeID, initialPost); err != nil {
+		b.log.Error("failed to save initial post", zap.Error(err))
 	}
 
 	metrics.PostDuration.Set(float64(time.Since(startTime).Nanoseconds()))
 	public.PostSeconds.Set(float64(time.Since(startTime)))
 	b.log.Info("created the initial post")
+	return post, postInfo, nil
+}
 
-	return nipost.AddInitialPost(b.localDB, nodeID, initialPost)
+// Obtain certificates for the poets.
+// We want to certify immediately after the startup or creating the initial POST
+// to avoid all nodes spamming the certifier at the same time when
+// submitting to the poets.
+func (b *Builder) certifyPost(ctx context.Context, post *types.Post, meta *types.PostInfo, ch []byte) {
+	client := NewCertifierClient(b.log, post, meta, ch, WithCertifierClientConfig(b.certifierConfig.Client))
+	certifier := NewCertifier(b.localDB, b.log, client)
+	certifier.CertifyAll(ctx, b.poets)
+
+	b.smeshingMutex.Lock()
+	b.certifiers[meta.NodeID] = certifier
+	b.smeshingMutex.Unlock()
 }
 
-func (b *Builder) run(ctx context.Context, sig *signing.EdSigner) {
-	defer b.log.Info("atx builder stopped")
+func (b *Builder) obtainPostFromLastAtx(
+	ctx context.Context,
+	nodeId types.NodeID,
+) (*types.Post, *types.PostInfo, []byte, error) {
+	atxid, err := atxs.GetLastIDByNodeID(b.cdb, nodeId)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("no existing ATX found: %w", err)
+	}
+	atx, err := b.cdb.GetFullAtx(atxid)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to retrieve ATX: %w", err)
+	}
+	if atx.NIPost == nil {
+		return nil, nil, nil, errors.New("no NIPoST found in last ATX")
+	}
+	if atx.CommitmentATX == nil {
+		if commitmentAtx, err := atxs.CommitmentATX(b.cdb, nodeId); err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to retrieve commitment ATX: %w", err)
+		} else {
+			atx.CommitmentATX = &commitmentAtx
+		}
+	}
+	if atx.VRFNonce == nil {
+		if nonce, err := atxs.VRFNonce(b.cdb, nodeId, b.layerClock.CurrentLayer().GetEpoch()); err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to retrieve VRF nonce: %w", err)
+		} else {
+			atx.VRFNonce = &nonce
+		}
+	}
 
+	b.log.Info("found POST in an existing ATX", zap.String("atx_id", atxid.Hash32().ShortString()))
+	meta := &types.PostInfo{
+		NodeID:        nodeId,
+		CommitmentATX: *atx.CommitmentATX,
+		Nonce:         atx.VRFNonce,
+		NumUnits:      atx.NumUnits,
+		LabelsPerUnit: atx.NIPost.PostMetadata.LabelsPerUnit,
+	}
+
+	return atx.NIPost.Post, meta, atx.NIPost.PostMetadata.Challenge, nil
+}
+
+func (b *Builder) obtainPost(ctx context.Context, nodeID types.NodeID) (*types.Post, *types.PostInfo, []byte, error) {
```

Comment:
Why not return a single object, we already have one that represents both `types.Post` and `types.PostInfo`:

```suggestion
func (b *Builder) obtainPost(ctx context.Context, nodeID types.NodeID) (*nipost.Post, error) {
```